Overhaul for the parser module

zhubonan opened the pull request:
**Please check the applicable boxes, thank you**:

I, the author consider this PR
 - [x] ready to be reviewed and merged (as soon as tests have passed)
 - [ ] work in progress (early feedback welcome)

## Interactions with issues / other PRs

*type "#" followed by search words to find issues / PRs*

fixes:

blocks:

is blocked by:

None of the above but is still related to the following:

## Description

This PR change the parser interface with the aim to make it more simple with the cost of reduced generalizability, but I think the latter is less important now as the community did not go into the code-sharing route among different plugins. 
The functionality of *quantity*, *settings* and *node composer* are moved into a single sub-class of the `Parser`. 

The logic of the parsing goes roughly as this:

1. Read each file with the content parser (content parsers are mostly not changed)
2. Collect the information into a dictionary (`VaspParser.quantities_each`)
3. Check required quantities and remove excluded quantities
4. Compose the output nodes

The default strategy is to parse all quantities except those that are excluded. The user may *include/exclude* quantities and nodes to modify based on the default setting. 

Another change is that now the common information are saved in the `misc` node as dictionary, including the forces and stress as they are typically not large arrays (size in the order of ~100 along the long dimension). 
Instead of saving most property as a separate node, they are not saved under the `arrays` output. The motivation here is to reduce the number of output nodes as `ArrayData` is capable of saving multiple arrays (which are serialized into numpy arrays in the repository). The exceptions are band structure (`bands`) and `dos` as they are considered as common DFT output. This also makes the parser code simpler to read and extend. 

The changes to the existing `misc` output is minimum and I expect a certain backward compatibiltiy, but this is certainly a breaking change. 

In addition, Pydantic is used to verify the `parser_settings` and provide default values. The `OptionalContainer` is now a sub-claass of the `BaseModel` (the previous implementation was a bit of reinventing the wheel...). This also allows checking of the settings at the *submission time* and alow IDE autocompletion. 

This PR includes the changes of #720

zhubonan committed:
Improved the BuilderUpdater interface

New apply_preset method all a preset to be used to generate the inputs
automatically, including resources, options and settings. The user
may store such defaults in a yaml file under ~/.aiida-vasp

zhubonan committed:
Add type hints

zhubonan committed:
Updated bands workchain and default preset

zhubonan committed:
Update band.py to make it work

zhubonan committed:
Initial commit to simplify the parser system

zhubonan committed:
Fix the convergence workchain

zhubonan committed:
Use pydantic for handling control options

Use pydantic for converting between python object and dictionary
used for controlling workchains. Rewrite OptionContainer to be
a sub-class of BaseModel,  retaining all previous functionalities.

zhubonan committed:
Fix and make aiida_validate work

zhubonan committed:
Move and options to the option holder module

zhubonan committed:
Merge branch 'inputset-improve' into parser-overhaul

zhubonan committed:
Update test and code

zhubonan committed:
Update and modify tests

zhubonan committed:
Update tests

zhubonan committed:
Add more test and make error composer work

zhubonan committed:
Make tests work for calcs/test_vasp and test_new_parser.py

zhubonan committed:
Update tests

zhubonan committed:
Temporary disable none-working tests

zhubonan committed:
Update parser settings used by workchains

zhubonan committed:
fix relax workchain

zhubonan committed:
Fix tests

zhubonan committed:
Add NEB parser

zhubonan committed:
Remove old parser code

atztogo commented on pyproject.toml:
```diff
@@ -37,7 +37,9 @@ dependencies = [
     "seekpath",
+    "pydantic",
     "packaging",
+    "pydantic",
```
Two pydantic lines.

atztogo commented on pyproject.toml:
```diff
@@ -60,7 +62,9 @@ docs = [
     "sphinx-rtd-theme",
     "sphinxcontrib-apidoc",
     "sphinxext.remoteliteralinclude",
+    "myst-parser"
```
Is this for moving to markdown?

zhubonan replied:
That is the plan. But it will be in a different PR. I should revert the changes.

atztogo replied:
No need to revert.

atztogo commented on src/aiida_vasp/calcs/neb.py:
```diff
@@ -51,23 +51,23 @@ def define(cls, spec):
         spec.input(
             'parameters',
-            valid_type=get_data_class('core.dict'),
+            valid_type=orm.Dict,
```
Have all `get_data_class` been changed to `orm.AiiDAType`?

zhubonan replied:
No yet, I should change them.

zhubonan replied:
Good call, there are still quite a few left. I will change them.

atztogo replied:
Thank you. It can be done by another PR.

atztogo commented:
> Another change is that now the common information are saved in the misc node as dictionary
> I expect a certain backward compatibiltiy, but this is certainly a breaking change.

This is a big change to me. Could you write a list of quantities that are moved to `misc`?

> The default strategy is to parse all quantities except those that are excluded. The user may include/exclude quantities and nodes to modify based on the default setting.

What are the default quantities and how are they included or excluded?

zhubonan commented:
> Another change is that now the common information are saved in the misc node as dictionary
> I expect a certain backward compatibiltiy, but this is certainly a breaking change.
> This is a big change to me. Could you write a list of quantities that are moved to misc?

The quantities moved to misc that are previously standalone nodes are: *forces*, *stress*, *site_magnetization*. You can see the full list below. I can certainly add options to maintain the old behaviour if needed. 

> What are the default quantities and how are they included or excluded?

All quantities that goes into misc now are:

```python
[    'total_energies',
    'maximum_stress',
    'maximum_force',
    'notifications',
    'run_status',
    'run_stats',
    'version',
    'forces',
    'stress',
    'site_magnetization',
    'band_properties',
    'elastic_moduli',
    'symmetries',
    'fermi_level',
    'band_properties',
    'magnetization',]
```

And for `arrays`: 

```python
('projectors', 'born_charges', 'dielectrics', 'hessian', 'dynmat', 'energies')
```

but the followings quantities are excluded by default:

```python
(
    'energies',
    'chgcar',
    'wavecar',
    'projectors',
    'charge_density',
    'magnetization_density',
    'elastic_moduli',
    'symmetries',
)
```

and the following nodes are excluded by default:

```python
['bands', 'dos', 'kpoints', 'trajectory']
```

The reason for excluding quantity/nodes is that the most common kind of calculation such as single point and relaxation does generate them, but they are not needed/not converged in most cases. In other words, they are needed only for calculations of specific properties.

To include a quantity/node:

```python
{'parser_settings': {'include_quantity': ['energies'], 'include_node': ['dos'], 'exclude_quantity': ['forces']}}
```

atztogo commented:
> And for arrays:

Do you mean the following quantities are stored in Arrays?

> `band_properties`

This name is not very nice. This seems rather band-gap information.

zhubonan commented:
> > And for arrays:
> 
> Do you mean the following quantities are stored in Arrays?

Yes, they are stored in an output node with the link name `arrays`. Some of these quantities have multiple arrays, the names are suffixed to the name of the quantity. For example,  `arraynode.get_array('energies_energy_extrapolated')`.

> > `band_properties`
> 
> This name is not very nice. This seems rather band-gap information.

Okay, thanks for the suggestion!

zhubonan commented:
> Yes, they are stored in an output node with the link name arrays. Some of these quantities have multiple arrays, the names are suffixed to the name of the quantity. For example, arraynode.get_array('energies_energy_extrapolated').

Maybe we should keep some as separate nodes. For example, sending `born_charges` as input phonopy workchain makes sense than a whole `arrays` node containing lots of other quantities.

zhubonan committed:
Remove get_data_class

Use of this function is no longer needed as aiida.orm can be imported
without loading the profile.

zhubonan committed:
Remove get_data_class and get_data_node

zhubonan committed:
Switch to standalone array nodes

Move individual properties to separate nodes

zhubonan commented:
@atztogo I have restored the old behaviour that array-like outputs will be stored in separate ArrayData nodes. The `arrays` node only collects miscellaneous outputs that are large arrays. 

I think this PR is ready to be merged

atztogo commented:
@zhubonan, do you think this will be released as v4?

zhubonan commented:
Yes, definite should be a v4 release. My plan is to put in a few more improvements and add the documentation, then release as v4 in a week or two. 
By the way, are you happy with switching the MyST markdown/ipython notebook (myst_nb) for documentations?

atztogo commented:
> are you happy with switching the MyST 

Sure!
